base/iterators.jl

```diff
@@ -1700,6 +1701,9 @@ function _nth(::IteratorSize, itr, n::Integer)
     y === nothing && throw(BoundsError(itr, n))
     y[1]
 end
+
+_nth(::Union{HasShape, HasLength}, z::Zip, n::Integer) = Base.map(nth(n), z.is)
```

Comment:
What about `SizeUnknown`?

Reply:
it will hit the generic fallback above that just calls `iterate` n times

Reply:
Great catch.  I didn't consider that case.  This does indeed fail on iterators of unknown size.  For example,
```julia
Iterators.filter(x -> x > 0, -5:5)
```
As humans, it's obvious that this is synonymous to `1:5`, but it doesn't implement `length`.

We might need use `_zip_lengths_finite_equal` (we explored using internal `Iterator` length functions in #58922 but initially decided against it).  Do you think this is sufficient (CC @adienes)?
```julia
function last(z::Zip)
    n = last(_zip_lengths_finite_equal(z.is))
    return nth(z, n)
end
```
Is it worth keeping the `length` implementation for iterators whose size is known?

Reply:
I'm not sure how `last` can know the length of the `Zip` if all constituent iterators have `SizeUnknown`.  Do we explicitly error in this case?

`nth` supports iterators whose size are unknown.  It would be nice to support it in `last` but I don't see any efficient way to get the length.

Reply:
`_zip_lengths_finite_equal` seems to handle the case where at least one of the constituent iterators have a known length, which our current implementation does not do:
```julia-repl
julia> z = zip(1:3, Iterators.filter(x -> x > 0, -5:5))
zip(1:3, Base.Iterators.Filter{var"#2#3", UnitRange{Int64}}(var"#2#3"(), -5:5))

julia> length(z)
ERROR: MethodError: no method matching length(::Base.Iterators.Filter{var"#2#3", UnitRange{Int64}})

julia> Iterators._zip_lengths_finite_equal(z.is)
(false, 3)
```

Reply:
```suggestion
_nth(::IteratorSize, z::Zip, n::Integer) = Base.map(nth(n), z.is)
```
and defer to the component iterators.

Reply:
Implemented in da76f8ce92 :slightly_smiling_face: